fdbserver/DDRelocationQueue.actor.cpp

```diff
@@ -2030,34 +2160,179 @@ ACTOR Future<bool> rebalanceReadLoad(DDQueue* self,
+	RelocateReason relocateReason = selectRelocateReason(self,
+	                                                     traceEvent,
+	                                                     moveReason,
+	                                                     reply.minReadLoad,
+	                                                     reply.maxReadLoad,
+	                                                     metrics.bytesReadPerKSecond,
+	                                                     srcLoad,
+	                                                     shard);
+
+	if (relocateReason == RelocateReason::OTHER) {
+		// do nothing
+		traceEvent->detail("SkipReason", "NotMoveShardWithMultipleTeams");
+		return false;
+	} else if (relocateReason == RelocateReason::MOVE_SHARD &&
+	           metrics.bytesReadPerKSecond > (srcLoad - destLoad) * SERVER_KNOBS->READ_REBALANCE_MAX_SHARD_FRAC) {
+		traceEvent->detail("SkipReason", "NotMoveSuperHotShard");
+		return false;
+	} else {
+		traceEvent->detail("RelocateReason", relocateReason.toString()).detail("ServerSize", sourceTeam->size());
 
-			self->serverCounter.increaseForTeam(
-			    serverIds, RelocateReason::REBALANCE_READ, DDQueue::ServerCounter::ProposedSource);
-			return true;
+		for (int i = 0; i < shards.size(); i++) {
+			if (shard == shards[i]) {
+				UID traceId = deterministicRandom()->randomUniqueID();
+				self->output.send(RelocateShard(shard, moveReason, relocateReason, traceId));
+				traceEvent->detail("TraceId", traceId);
+
+				auto serverIds = sourceTeam->getServerIDs();
+				self->updateLastAsSource(serverIds);
+
+				self->serverCounter.increaseForTeam(serverIds, relocateReason, DDQueue::ServerCounter::ProposedSource);
+				return true;
+			}
 		}
+		traceEvent->detail("SkipReason", "ShardNotPresent");
+		return false;
 	}
-	traceEvent->detail("SkipReason", "ShardNotPresent");
-	return false;
+}
+
+ACTOR Future<bool> rebalanceByReduceReplicas(DDQueue* self,
+                                             Optional<Reference<IDataDistributionTeam>> _randomTeam,
+                                             DataMovementReason moveReason,
+                                             bool primary) {
+	if (!SERVER_KNOBS->DYNAMIC_REPLICATION_ENABLED) {
+		return false;
+	} else if (!_randomTeam.present()) {
+		return false;
+	} else {
+		state Reference<IDataDistributionTeam> randomTeam = _randomTeam.get();
+
+		state std::vector<KeyRange> shards = self->shardsAffectedByTeamFailure->getShardsFor(
+		    ShardsAffectedByTeamFailure::Team(randomTeam->getServerIDs(), primary));
+		state TraceEvent traceEvent("RebalanceByReduceShard");
+
+		if (shards.size() < 1) {
+			traceEvent.detail("SkipReason", "NoShardOnSource");
+			return false;
+		}
+
+		int bottomK = std::min(int(0.1 * shards.size()), SERVER_KNOBS->READ_REBALANCE_SHARD_TOPK);
+		if (self->timeThrottle(randomTeam->getServerIDs())) {
+			traceEvent.detail("SkipReason", "SourceTeamThrottle");
+			return false;
+		}
+
```

Comment:
I wonder how much cost does reducing replica on the source servers? Maybe the time throttle is unnecessary for replication reduction